### ISSUE TITLE
showmodal issue and autoresize LCL bug

### ISSUE THREAD

**Issue by mgrinzPlayer** [cheat-engine/cheat-engine#115]
**autosize LCL bug**
About autoresize, I made example project and submitted new issue on Lazarus bugtracker.
http://bugs.freepascal.org/view.php?id=30826

Currently I'm below patch. Minimizing "dissect code" and "trainer generator" works normally.
```patch
Index: customform.inc
===================================================================
--- customform.inc	(wersja 53269)
+++ customform.inc	(kopia robocza)
@@ -712,6 +712,9 @@
     end;
   end;
 
+  if (Message.SizeType xor SIZE_SourceIsInterface)=SIZE_MINIMIZED then
+    Resizing(wsMinimized);
+
   inherited WMSize(Message);
 
   if (Message.SizeType and not SIZE_SourceIsInterface) = SIZE_RESTORED then
@@ -2226,6 +2229,7 @@
 ------------------------------------------------------------------------------}
 function TCustomForm.AutoSizeDelayedHandle: Boolean;
 begin
+  if WindowState=wsMinimized then exit(true);
   if (Parent<>nil) or (ParentWindow<>0) then
     // this form is inlined / embedded it works like a normal TWinControl
     Result:=inherited AutoSizeDelayedHandle
```

This bug can also cause created forms (UDF1) to freeze. I attached example below
[form crash.ct.zip](https://github.com/cheat-engine/cheat-engine/files/563017/form.crash.ct.zip)

**Comment by mgrinzPlayer** [comment]
**showmodal issue**
Looks like this modification
https://github.com/cheat-engine/cheat-engine/commit/2d8d22b07e2beefde99e70eba7504d0999babb77

was supposed to revert changes made earlier
https://github.com/cheat-engine/cheat-engine/commit/b423ca7888a2dc8615f75d7aea5fc8e42493e3b0

But, something is wrong. For example, TfrmStructures2ElementInfo still brings the mainform instead of TfrmStructures2.

I restored PopupMode = pmAuto inside frmStructures2ElementInfo.lfm and it works better.

The same problem is with commenting a disassembled line inside Memory View or when you open editor preferences. But, restoring for all modal forms may not be a good option.


I found out that there is something about this issue in rev.51686
```patch
Index: customform.inc
===================================================================
--- customform.inc	(wersja 51685)
+++ customform.inc	(wersja 51686)
@@ -1466,15 +1466,16 @@
      (PopupMode in [pmAuto, pmExplicit]) // set WndParent of non-modal windows only for pmAuto, pmExplicit
   then
   begin
-    case PopupMode of
-      pmAuto:
-      begin
-        Result := Screen.ActiveForm;
-        if (Result<>nil) and (Result.FormStyle = fsSplash) then // ignore fsSplash
-          Result := nil;
-      end;
-      pmExplicit: Result := PopupParent;
-    end;
+    if (PopupMode = pmAuto)
+    or ((PopupMode = pmNone) and (fsModal in FormState)) then
+    begin
+      Result := Screen.ActiveForm;
+      if (Result<>nil) and (Result.FormStyle = fsSplash) then // ignore fsSplash
+        Result := nil;
+    end else
+    if (PopupMode = pmExplicit) then
+      Result := PopupParent;
+
     if (Result = nil) or not Result.HandleAllocated then
       Result := Application.MainForm;
   end;
@@ -2875,6 +2876,11 @@
   Application.ModalStarted;
   try
     Include(FFormState, fsModal);
+    if PopupMode = pmNone then
+    begin
+      RecreateWnd(Self);
+      HandleNeeded;
+    end;
     ActiveWindow := GetActiveWindow;
     SavedFocusState := SaveFocusState;
     SavedCursor := Screen.Cursor;

```

**Comment by mgrinzPlayer** [comment]
If it is not a big issue, I recommend:
- attached patch for LCL Lazarus,
[LCL.patch.zip](https://github.com/cheat-engine/cheat-engine/files/562964/LCL.patch.zip)

- accept pull requests (maybe except RLE thing, it's your call)
- rebuild Lazarus, rebuild CE, release **installer 6.6.0.2**

We still have time for your usual "Post release fixes max 30 days if a HUGE bug"

**Comment by cheat-engine** [comment]
Adjusting the popup mode for some specific dialogs might work too

Anyhow, under HUGE bugs I count CE crashing completely for ordinary tasks and things not working like the Alloc auto assembler command.
A few popups being annoying falls in the minor category
The all scan not is slightly more, but judging by the number of people reporting this issue(1), I don't think it's a huge problem

And I can't make a new build at the moment anyhow. (on vacation where I don't have my code signing dongle with me)

**Comment by mgrinzPlayer** [comment]
If someone decides to use a form with few autosize panels and form is also autosize, it will crash.

It is not that uncommon thing. I mean: autosize window. Here a real life example:
I made this trainer some time ago with CE6.4:
http://forum.cheatengine.org/viewtopic.php?p=5525146#5525146

go to TMP\cetrainers\CETXXXXX.tmp\extracted\autorun there is options.lua

If you open it and go to Options:buttonClicked function, you will see how "trainer settings" form is created. Here I combined my options.lua and small example into one file: [settings form crash.cetrainer.zip](https://github.com/cheat-engine/cheat-engine/files/563193/settings.form.crash.cetrainer.zip)


So, we can say that "backward compatibility" is also broken.
